Update collection loader for Python 3.10

s-hertel opened the pull request:
##### SUMMARY
Fixes #74660

Implement find_spec and exec_module to remove reliance on deprecated methods find_module and load_module.

##### ISSUE TYPE
- Feature Pull Request

s-hertel committed:
Implement find_spec and exec_module to remove reliance on deprecated methods in the collection loader

ci_complete

sivel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -224,6 +230,52 @@ def find_module(self, fullname, path=None):
             # TODO: log attempt to load context
             return None
 
+    def find_spec(self, fullname, path, target=None):
```
I think I'd like to see this consolidated with `find_module`. They are about 99% the same.  My recommendation is to drop everything from `find_module` and then do:

```python
    def find_module(self, fullname, path=None):
        spec = self.find_spec(fullname, path=path)
        if spec:
            return spec.loader
        return None
```

nitzmahone replied:
Yep

s-hertel replied:
Done

sivel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -285,6 +337,33 @@ def find_module(self, fullname, path=None):
                 return pkgutil.ImpImporter(self._pathctx).find_module(fullname)
 
+    def find_spec(self, fullname, target=None):
+        split_name = fullname.split('.')
```
This one is a bit more complicated it seems with the py2 branches, but I think we can still manage combine them.

nitzmahone replied:
and yep.

sivel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -377,6 +456,29 @@ def _module_file_from_path(leaf_name, path):
+        if spec.submodule_search_locations is not None:
+            module.__path__ = spec.submodule_search_locations
+
+        sys.modules[spec.name] = module
```
I don't believe `create_module` should be adding to `sys.modules`.  The `importlib` documentation isn't clear here, but reading pep489 leads me to believe this should only happen during the `Module Execution Phase`.

I suppose it doesn't really matter in our context, since I doubt it's being used independently, but for correctness I believe it should be moved to `exec_module`.

`create_module` should only create the module shell, and the code isn't actually loaded until `exec_module`.  Which means that `create_module` shouldn't be loading any resources from the collection yet.

This creates a problem with `_AnsibleCollectionPkgLoader.create_module` though, since I think it is doing too much at that point.  I think `_AnsibleCollectionPkgLoader.create_module` probably needs removed and it's logic moved into an `exec_module`, or at minimum `AnsibleCollectionConfig.on_collection_load.fire` moved out of `create_module`, otherwise we get a recursion error.

I'd like to see what @nitzmahone thinks here.

nitzmahone replied:
I can't think of any reason off the top of my head that we'd need to customize the module object creation, so just having create_module return `None` *should* be fine. 

IIRC `exec_module` is literally just about populating the module (already created by create_module or the base import machinery) from the code; the 3.x import machinery takes care of the transactional insertion/removal from `sys.modules` on fresh imports and reloads, so we should probably never mess with that in the 3.x loader code at all. It's been awhile, but I think it basically just calls create_module (and does the stock creation if we didn't return something), provisionally inserts the module to `sys.modules`, calls exec_module under an exception handler, and removes the module on an error if it inserted it.

s-hertel replied:
I removed all of the extra sys.modules stuff, except from the redirect loader (which is maybe wrong?). There's a comment in importlib reload() that seems like it supports exec_module modifying sys.modules at least, but maybe I'm using it as intended.
```
        _bootstrap._exec(spec, module)
        # The module may have replaced itself in sys.modules!
        return sys.modules[name]
```

sivel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -377,6 +456,29 @@ def _module_file_from_path(leaf_name, path):
+            return self._redirect_module
+
+        module = ModuleType(spec.name)
+        module.__loader__ = spec.loader
```
Reading pep489 I also see this in regards to the creation phase:

```
This function is not responsible for setting import-related attributes specified in PEP 451 [1] (such as __name__ or __loader__) on the new module.
```

pep451 (https://www.python.org/dev/peps/pep-0451/#attributes) says:

> The reverse of this mapping describes how the import machinery sets the module attributes right before calling exec_module().

So I'm thinking we shouldn't be doing most of this manipulation of `module` other than just creating the module object.

Sorry, I'm beginning to ramble as I think this through...

But based on this and other comments, I question whether we need `create_module` at all.  Or rather, we just have it return `None`.

nitzmahone replied:
Yep

s-hertel replied:
Fixed

nitzmahone reviewed (changes requested):
Added comments to sivel's; the other piece of this that looks plausible is actually just getting rid of ansible-test's copy of the collection loader. We assumed it wouldn't be very feasible to have a combo 2.x/3.x loader, but it kinda looks like it might be, so if this all works, we should probably (either as part of this PR or right after) get rid of the second copy altogether.

webknjaz commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -43,6 +43,12 @@ def import_module(name):
 
+try:
+    from importlib.util import spec_from_loader
+    PY3_SPEC_LOADER = True
```
Could you put this into the else-block?

s-hertel replied:
Done

webknjaz commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -676,6 +816,28 @@ def __init__(self, fullname, path_list):
             raise ImportError('not redirected, go ask path_hook')
 
+    def get_filename(self, fullname):
+        return None
```
Should this be something like `<synthetic package/module>`? I know that `eval()`/`python -c` does something like this.

s-hertel replied:
Hm... `'<ansible_synthetic_redirection>'`? Feel free to bikeshed

webknjaz replied:
I don't have any preference. The only thing is that it probably shouldn't use underscores instead of plain old whitespace. It's not an identifier, after all. But I won't insist on changing this.

s-hertel replied:
Okay, I'll fix that.

s-hertel committed:
Move module execution to exec_module

Remove extra sys.modules handling

Use default module initialization by returning None from loader.create_module

Refactor

ci_complete

s-hertel committed:
Remove ansible-test's copy of the collection loader

ci_complete

s-hertel committed:
Fix metaclass for Python 2.x

ci_complete

webknjaz commented on lib/ansible/utils/collection_loader/_collection_config.py:
```diff
@@ -101,6 +102,10 @@ def _require_finder(cls):
-# concrete class of our metaclass type that defines the class properties we want
-class AnsibleCollectionConfig(metaclass=_AnsibleCollectionConfig):
-    pass
+if PY3:
```
`six` has a `@with_metaclass` decorator IIRC so that you don't need to do this if-else dance. LMGTFY

webknjaz replied:
Oh, it's `@add_metaclass` that is a decorator and `with_metaclass()` is a function. FWIW here's the doc on both: https://six.readthedocs.io/#six.with_metaclass.

s-hertel replied:
:100: much better

s-hertel replied:
Fixed, thanks for the pointer

webknjaz commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -211,19 +225,26 @@ def find_module(self, fullname, path=None):
         try:
```
Can this be refactored a bit? Having a big try-block is an antipattern. The loader factory selection will never cause an import error so it could be moved out and then the only thing that may cause this exception would be in try-except.
```python
if ...:
    initialize_loader = ...
elif ...:
    initialize_loader = ...
else:
    initialize_loader = _AnsibleCollectionLoader

try:
    loader = initialize_loader(fullname=fullname, path_list=path)
except ImportError:
    return None
```

webknjaz commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -256,9 +277,21 @@ def find_module(self, fullname, path=None):
+        # call py2's internal loader
+        if toplevel_pkg != 'ansible_collections' and not PY3:
+            return pkgutil.ImpImporter(self._pathctx).find_module(fullname)
+        else:
```
`else` is unnecessary here, you could drop it to decent the block. Not sure why this pylint rule is disabled :man_shrugging:

s-hertel replied:
fixed

s-hertel committed:
Fix Py2/Py3 syntax compatibility

s-hertel committed:
Refactor

ci_complete

s-hertel committed:
update collection_loader comments

ci_complete

s-hertel committed:
simplify find_module

ci_complete

s-hertel committed:
Fix Py2 compatibility - don't get loader from nonexistent spec

Remove unnecessary PY3 checking

sivel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -531,6 +567,42 @@ def _validate_final(self):
             self._subpackage_search_paths = [self._subpackage_search_paths[0]]
 
+    def exec_module(self, module):
+        if not _meta_yml_to_dict:
```
Since this duplicates much of `load_module`, maybe we could just call the 3 step loader from `load_module` so that we eliminate the duplication?

s-hertel replied:
I refactored those two methods to both use a new method `_load_module`, similar to what I did for find_module/find_spec in the commit before it.

s-hertel committed:
Refactor common code in load_module and exec_module

ci_complete

s-hertel committed:
tidy diff

ci_complete

s-hertel commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -43,6 +38,11 @@ def import_module(name):
+try:
+    from importlib.util import spec_from_loader
+except ImportError:
+    pass
```
It shouldn't ever happen, but maybe I should set spec_from_loader to None here, and raise a ValueError in find_spec if it's None.

nitzmahone replied:
I'm actually a little surprised pylint doesn't complain about the possibly undefined name here, but if it genuinely doesn't, we're basically getting the behavior we want for free the way the code is now (a `NameError` on `spec_from_loader` if someone calls `find_spec`, which is probably more useful to anyone that cares than the extra effort required for more explicit error handling. So I'd vote for "leave as is" if pylint's not complaining.

mattclay commented:
@s-hertel The Python version filter in `ansible-test` should be updated to include the collection loader in the tests for target-only Python versions. That can be done by adding `lib/ansible/utils/collection_loader` to the list of paths here: https://github.com/ansible/ansible/blob/fe77bc9e3c44d3b340f258397b7081de893c2680/test/lib/ansible_test/_internal/commands/sanity/__init__.py#L774

The main effect of that change will be to include the collection loader in the `compile` tests for all Python versions.

samccann commented:
does it warrant a changelog fragment??

s-hertel committed:
Include collection_loader in target paths for 'compile' sanity test

s-hertel committed:
add changelog

s-hertel commented:
@samccann Probably! Added one.

samccann reviewed (approved):
changelog fragment lgtm :-)

nitzmahone commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -277,13 +289,30 @@ def find_module(self, fullname, path=None):
+    def find_module(self, fullname, path=None):
+        # we ignore the passed in path here- use what we got from the path hook init
+        split_name = fullname.split('.')
+        toplevel_pkg = split_name[0]
```
these two lines are vestigial and can be removed, yeah?

s-hertel replied:
Removed, good catch

nitzmahone commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -277,13 +289,30 @@ def find_module(self, fullname, path=None):
+
+        finder = self._get_finder(fullname)
+        if finder is not None:
+            return finder.find_module(fullname, path=[self._pathctx])
```
It effectively works this way anyway, but probably best to add an explicit `else: return None`, just in case any other logic were to be added to this function in the future.

s-hertel replied:
Added `else: return None` everywhere that was doing it implicitly

nitzmahone commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -377,6 +406,17 @@ def _module_file_from_path(leaf_name, path):
         return module_path, has_code, package_path
 
+    def exec_module(self, module):
+        # execute the module's code in its namespace
```
This should probably no-op in the `if self._redirect_module:` case, otherwise, depending on what code we find laying around, we're potentially re-initializing an existing module or trashing it with empty/wrong code.

s-hertel replied:
Fixed.

nitzmahone commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -377,6 +406,17 @@ def _module_file_from_path(leaf_name, path):
+    def create_module(self, spec):
+        # short-circuit redirect; we've already imported the redirected module, so just alias it and return it
+        if self._redirect_module:
+            return self._redirect_module
```
another case where it functions properly as-is, but an explicit `else: return None` is probably apropos just in case.

nitzmahone commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -676,6 +725,20 @@ def __init__(self, fullname, path_list):
         if not self._redirect:
             raise ImportError('not redirected, go ask path_hook')
 
+    def get_filename(self, fullname):
```
This doesn't seem right at a glance; IIRC the import machinery calls this after load/exec_module to stuff the `__file__` attr on the module- this would overwrite a legit filesystem `__file__` entry from the import we did with this value. Do we know for sure that this is needed here?

s-hertel replied:
Good catch. I've removed it and added ci_complete to the latest commit. I don't see why I thought defining the method was necessary.

nitzmahone reviewed (changes requested):
Looking good! A few extraneous lines, a couple of recommended explicit returns, and a couple of corner cases to talk through, but I think this is pretty much ready to go.

s-hertel committed:
Add "return None" instead of doing it implicitly

Remove get_filename

short-circuit exec_module if it's a redirect

ci_complete

webknjaz commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -277,13 +291,31 @@ def find_module(self, fullname, path=None):
+        finder = self._get_finder(fullname)
+        if finder is not None:
+            if toplevel_pkg == 'ansible_collections':
+                return finder.find_spec(fullname, path=[self._pathctx])
             else:
-                # call py2's internal loader
-                return pkgutil.ImpImporter(self._pathctx).find_module(fullname)
+                return finder.find_spec(fullname)
+        else:
+            return None
```
nitpick: I think this could have less nesting.
```suggestion
        if finder is None:
            return None

        if toplevel_pkg == 'ansible_collections':
            return finder.find_spec(fullname, path=[self._pathctx])

        return finder.find_spec(fullname)
```

webknjaz commented on lib/ansible/utils/collection_loader/_collection_finder.py:
```diff
@@ -207,23 +205,40 @@ def find_module(self, fullname, path=None):
+
+    def find_spec(self, fullname, path, target=None):
+        loader = self._get_loader(fullname, path)
+        if loader:
+            spec = spec_from_loader(fullname, loader)
+            if spec is not None and hasattr(loader, '_subpackage_search_paths'):
+                spec.submodule_search_locations = loader._subpackage_search_paths
+            return spec
+        else:
+            return None
```
nit: this is a good place to use a "guard expression" 
```suggestion
        if not loader:
            return None

        spec = spec_from_loader(fullname, loader)
        if spec is not None and hasattr(loader, '_subpackage_search_paths'):
            spec.submodule_search_locations = loader._subpackage_search_paths

        return spec
```

nitzmahone reviewed (approved):
Looks great- thanks!

s-hertel commented:
Thanks for all the reviews!

webknjaz reviewed (commented):
I forgot to publish this draft review. So posting it post-merge for history. In case somebody will want to refactor these things.

felixfontein commented:
This PR seems to cause some problems in some community.general unit tests for Python 3.7 to 3.9: https://dev.azure.com/ansible/community.general/_build/results?buildId=30854&view=logs&j=83b878b7-b763-5cdf-bf1e-fd8bdd476fa6&t=450760e6-4f51-5fd0-aaa0-1d38d376b06c

felixfontein commented:
```
_______________ ERROR collecting cloud/linode/test_linode_v4.py ________________
02:10 /usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py:2194: in _handle_ns
02:10     loader = importer.find_spec(packageName).loader
02:10 E   AttributeError: 'NoneType' object has no attribute 'loader'
02:10 
02:10 During handling of the above exception, another exception occurred:
02:10 tests/unit/plugins/modules/cloud/linode/test_linode_v4.py:10: in <module>
02:10     linode_apiv4 = pytest.importorskip('linode_api4')
02:10 /usr/local/lib/python3.9/dist-packages/linode_api4/__init__.py:3: in <module>
02:10     from linode_api4.linode_client import LinodeClient
02:10 /usr/local/lib/python3.9/dist-packages/linode_api4/linode_client.py:7: in <module>
02:10     import pkg_resources
02:10 /usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py:3243: in <module>
02:10     def _initialize_master_working_set():
02:10 /usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py:3226: in _call_aside
02:10     f(*args, **kwargs)
02:10 /usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py:3268: in _initialize_master_working_set
02:10     tuple(
02:10 /usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py:3269: in <genexpr>
02:10     dist.activate(replace=False)
02:10 /usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py:2783: in activate
02:10     declare_namespace(pkg)
02:10 /usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py:2282: in declare_namespace
02:10     _handle_ns(packageName, path_item)
02:10 /usr/local/lib/python3.9/dist-packages/pkg_resources/__init__.py:2199: in _handle_ns
02:10     loader = importer.find_module(packageName)
02:10 /tmp/ansible-test-u4ejf2o8/ansible/utils/collection_loader/_collection_finder.py:303: in find_module
02:10     return finder.find_module(fullname, path=[self._pathctx])
02:10 E   TypeError: _find_module_shim() got an unexpected keyword argument 'path'
```

s-hertel commented:
@felixfontein I'm looking at this but don't have a fix yet. Thanks for spotting the issue. Running the tests locally fails during getting the metadata, so I'm not sure if my pokings around  going in the right direction (not sure if that's an additional bug or if I'm just running the tests wrong). Do you know if there is a way to open a PR against community.general to run CI against an open ansible/ansible PR?

felixfontein commented:
@s-hertel there isn't a "ready to use" way to do this; what I usually do is a) kick out almost everything in https://github.com/ansible-collections/community.general/blob/main/.azure-pipelines/azure-pipelines.yml so that only the tests I'm interested are running, and b) modify https://github.com/ansible-collections/community.general/blob/main/tests/utils/shippable/shippable.sh#L60 to install another branch from another user.

s-hertel commented:
@felixfontein Thanks much, good to know :smile: I have a local reproducer now (I was running the test without --docker).
